accumulator/invocation.go

```diff
@@ -0,0 +1,106 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package accumulator
+
+import (
+	"bytes"
+	"time"
+
+	"github.com/tidwall/gjson"
+	"go.elastic.co/fastjson"
+)
+
+// Invocation holds data for each function invocation and finalizes
+// the data when `platform.report` type log is received for the
+// specific invocation identified by request ID.
+type Invocation struct {
+	// RequestID is the id to identify the invocation.
+	RequestID string
+	// Timestamp is the time of the invocation.
+	Timestamp time.Time
+	// DeadlineMs is the function execution deadline.
+	DeadlineMs int64
+	// FunctionARN requested. Can be different in each invoke that
+	// executes the same version.
+	FunctionARN string
+	// TransactionID is the ID generated for a transaction for the
+	// current invocation. It is populated by the request from agent.
+	TransactionID string
+	// TraceID is the ID generated for a trace for the current invocation.
+	// It is populated by the request from agent.
+	TraceID string
+	// Metadata stripped data from the agent. Each line is represented as
+	// a seperate entry.
+	agentData [][]byte
+}
+
+// Finalize searches the agent data for an invocation to find the root txn
+// for the invocation. If root txn is not found then a new txn is created
+// with the payload submitted by the agent.
+func (inc *Invocation) Finalize(status string) {
+	if inc.TransactionID == "" {
+		return
+	}
+	if rootTxnIdx := inc.findRootTxn(); rootTxnIdx == -1 {
+		inc.createProxyTxn(status)
+	}
+}
+
+func (inc *Invocation) createProxyTxn(status string) int {
+	var w fastjson.Writer
+	w.RawString(`{"transaction":{"id":`)
+	w.String(inc.TransactionID)
+	w.RawString(`,"trace_id":`)
+	w.String(inc.TraceID)
+	w.RawString(`,"result":`)
+	w.String(status)
+	w.RawString("}}")
```

Comment:
Other fields I think we should be setting:
 - `name` (to what I'm not sure... I think ideally we would extract the payload and use it to set the name)
 - `type` ("request"? ideally we would set it according to the trigger type as described in https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-aws-lambda.md)
 - `timestamp` (to the invocation timestamp?)

Reply:
Updated the logic to use the fully encoded JSON for the partially created txn in the agent. All these fields (at least for the GoAgent) will be set.